### ISSUE TITLE
Closure using parent-scope vars incorrectly reports used-before-def after if statements

### ISSUE THREAD

**Issue by doublethefish** [python/mypy#15145]
```python
def outer(var: str) -> typing.Callable:
    def closure() -> None:
        if var == "":
            var = "function.__name__"
        print(var)

    return closure

```
**Bug Report**

Closures using variables from the parent-scope are reported as `used-before-def` when variables are used in if conditions

**To Reproduce**

1. Run mypy 1.2.0 against the following snippet
2. notice the `used-before-def` error erroneously reported

```python
def false_negative_if_first_statement(var: str) -> Callable:
    def closure() -> None:
        if var == "":  # <-- used-before-def
            var = "function.__name__"
        print(var)

    return closure


def false_negative_use_before_if(var: str) -> Callable:
    def closure() -> None:
        print(var)  # <-- used-before-def
        if var == "":  # <-- used-before-def
            var = "function.__name__"

    return closure


def expected_no_conditional(var: str) -> Callable:
    def closure() -> None:
        print(var)  # <-- no used-before-def

    return closure


def expected_conditional_no_if(var: str) -> Callable:
    def closure() -> None:
        print(var == "")  # <-- no used-before-def

    return closure
```

**Expected Behavior**

The use-cases are valid in all four examples above and shouldn't report `used-before-def`

**Actual Behavior**

```bash
error: Name "var" is used before definition  [used-before-def]
error: Name "var" is used before definition  [used-before-def]
error: Name "var" is used before definition  [used-before-def]
```
**Your Environment**

- Mypy version used: 1.2.0
- Mypy command-line flags: `python3 -m mypy`
- Mypy configuration options from `mypy.ini` (and other config files):
```ini
[mypy]

exclude = ./docs
disallow_untyped_defs=True
check_untyped_defs=True
show_error_codes=True
# disallow_typeddict_item=True
# disallow_untyped_calls=True
plugins = numpy.typing.mypy_plugin

```
- Python version used: Python 3.10.4

**Comment by erictraut** [comment]
I think mypy is correct here. The symbol `var` within the inner function `closure` is not the same symbol as `var` in the `outer` scope. They are separate variables. If you want it to reference the outer variable `var`, you need to add a `nonlocal var` statement within `closure`. As a demonstration of this, add the statement `outer("")()` at the end of the file and run it. This will result in an "UnboundLocalError: local variable 'var' referenced before assignment" error.

**Comment by JelleZijlstra** [comment]
Indeed.

**Comment by doublethefish** [comment]
Ah yes, we can demonstrate that it's not the if by making the var read-only in the closure. I should have thought to try that.

```python
def xxxxxx(var: str) -> Callable:
    def closure() -> None:
        if var == "":
            print(var)

    return closure
```

However, I would suggest one of two improvements:
1. as the variable is ghosted the error report or docs on the error report should hint at use of `nonlocal`
2. as the variable is ghosted it could be a separate issue with a better error description

But, thanks for the quick replies. I shall use `nonlocal` if/when I see this issue again.
